### PR TITLE
Public partition selection for Spark (#50)

### LINKED ISSUE #50: Implement filter_by_key in SparkRDDOperations
## Context

On [issue](https://github.com/OpenMined/PipelineDP/issues/4) `filter_by_key` method was implemented in [BeamOperations](https://github.com/OpenMined/PipelineDP/blob/e33a7c724179dba77b7646597f2af80e47ffe5a7/pipeline_dp/pipeline_operations.py#L118) and [LocalOperations](https://github.com/OpenMined/PipelineDP/blob/e33a7c724179dba77b7646597f2af80e47ffe5a7/pipeline_dp/pipeline_operations.py#L290).

## Goals
Implement `filter_by_key` in SparkRDDOperations with API:

`def filter_by_key(self, col, public_partitions)`

`col` is a collection (multimap) of **(partition_key, data)** (type of data is not important)
`public_partitions` is a collection of **(partition_key, )**,

`filter_by_key` returns `[record for record in collection if record[0] in public_partitions]`

1. If `public_partitions` is some Python collection then we can assume that memory is enough, Map operations is enough.
2. If `public_partitions` is RDD then join is required

## Open questions
Is it possible to unify more `filter_by_key` between `BeamOperations/SparkRDDOperations`? For example to come up with the join API in `PipelineOperations` and to implement `filter_by_key` using this API.

### PR DESCRIPTION
**Pull request by boyaryn**
## Description
Implementation of public partition selection for Spark processing engine.
Resolves #50 .

## Affected Dependencies
N/A

## How has this been tested?
Unit tests have been added to tests/pipeline_operations_test.py.

## Checklist
- [x] I have followed the [Contribution Guidelines](https://github.com/OpenMined/.github/blob/master/CONTRIBUTING.md) and [Code of Conduct](https://github.com/OpenMined/.github/blob/master/CODE_OF_CONDUCT.md)
- [x] I have commented my code following the [OpenMined Styleguide](https://github.com/OpenMined/.github/blob/master/STYLEGUIDE.md)
- [x] I have labeled this PR with the relevant [Type labels](https://github.com/OpenMined/.github/labels?q=Type%3A)
- [x] My changes are covered by tests

### CODE REVIEW

File: pipeline_dp/pipeline_operations.py [hunk 1]
```diff
@@ -42,7 +43,8 @@ def filter_by_key(self, col, public_partitions, stage_name: str):
 
         Args:
           col: collection with elements (partition_key, data).
-          public_partitions: collection of public partition keys.
+          public_partitions: collection of public partition keys,
```
Comment (dvadym) [inline pipeline_dp/pipeline_operations.py:46]: please rename `public_partitions` to `keys_to_keep` and in other `filter_by_key` implementations as well.

**Explanation**: we try to keep `PipelineOperations` are focused on pipeline operations not to DP. I missed that naming in previous code reviews.

**Review by dvadym — COMMENTED** [review]
Thanks it looks great! I left some comments

File: examples/movie_view_ratings.py [hunk 2]
```diff
@@ -126,7 +126,11 @@ def compute_on_spark():
     movie_views = sc.textFile(FLAGS.input_file) \
         .mapPartitions(parse_partition)
     pipeline_operations = pipeline_dp.SparkRDDOperations()
-    dp_result = calc_dp_rating_metrics(movie_views, pipeline_operations)
+    public_partitions = None
+    if FLAGS.public_partitions is not None:
```
[the remark this answers is not in the document]
Comment (boyaryn) [reply examples/movie_view_ratings.py:130]: Thanks, I forgot about DRY.

File: pipeline_dp/pipeline_operations.py [hunk 3]
```diff
@@ -223,8 +225,31 @@ def filter_by_key(self,
...
+                return rdd.filter(
+                    lambda x: x[0] in public_partitions
+                )
+
+            else:
+                filtering_rdd = public_partitions.map(
+                    lambda x: (x, None)
+                )
+                return rdd.join(filtering_rdd).map(
+                    # lambda partition, data: (partition, data[0])
```
[the remark this answers is not in the document]
Comment (boyaryn) [reply pipeline_dp/pipeline_operations.py:250]: Oh, I experimented with different ways to define a lamdba and forgot to delete an unsuccessful one.

File: pipeline_dp/pipeline_operations.py [hunk 4]
```diff
@@ -223,8 +225,31 @@ def filter_by_key(self,
                       public_partitions,
                       data_extractors,
                       stage_name: str = None):
-        NotImplementedError(
-            "filter_by_key is not implemented in SparkRDDOperations")
+
+        if public_partitions is None:
+            raise TypeError("Must provide a valid public_partitions")
+
+        else:
```
[the remark this answers is not in the document]
Comment (boyaryn) [reply pipeline_dp/pipeline_operations.py:232]: done

File: pipeline_dp/pipeline_operations.py (hunk 1 above)
Comment (boyaryn) [reply pipeline_dp/pipeline_operations.py:46]: changed

**Review by dvadym — APPROVED** [review]
Thanks!

**Pull request merged**